tests/frontend/mirror.py

```diff
@@ -807,3 +807,93 @@ def test_mirror_expand_project_and_toplevel_root(cli, tmpdir):
         # Success if the expanded %{project-root} is found
         assert foo_str in contents
         assert bar_str in contents
+
+
+# Test a simple SourceMirror implementation which reads
+# plugin configuration and behaves in the same way as default
+# mirrors but using data in the plugin configuration instead.
+#
+@pytest.mark.datafiles(DATA_DIR)
+@pytest.mark.usefixtures("datafiles")
+def test_source_mirror_plugin(cli, tmpdir):
+    output_file = os.path.join(str(tmpdir), "output.txt")
+    project_dir = str(tmpdir)
+    element_dir = os.path.join(project_dir, "elements")
+    os.makedirs(element_dir, exist_ok=True)
+    element_name = "test.bst"
+    element_path = os.path.join(element_dir, element_name)
+    element = generate_element(output_file)
+    _yaml.roundtrip_dump(element, element_path)
+
+    project_file = os.path.join(project_dir, "project.conf")
+    project = {
+        "name": "test",
+        "min-version": "2.0",
+        "element-path": "elements",
+        "aliases": {
+            "foo": "FOO/",
+            "bar": "BAR/",
+        },
+        "mirrors": [
+            {
+                "name": "middle-earth",
+                "kind": "mirror",
+                "aliases": {
+                    "foo": ["<invalid>"],
```

Comment:
^ This is what I am referring to. We shouldn't have to do this when implementing source mirrors.

Reply:
Just to clarify, this is not _necessary_ to do per se, while declaring aliases is still a requirement which may be lifted later in order to provide mirrors more generically without specific alias targeting.

To be clear, it is written this way in the _test case_ only to demonstrate that it is the `SourceMirror` in action, retrieving the information required to complete the URL in it's own plugin configuration rather than in the base mirror configuration.

Reply:
Can it be written any other way? BuildStream won't accept setting aliases to a list, it needs to be a dict. I haven't tested, but I assume it won't let me set the values to an empty list either (or would it?).

The end result is that, as a user, I need to set the `aliases` key to a bogus dictionary where only keys are taken from it, and those keys are repeated in the `config` below.

Reply:
This is just a test case, and the `"<invalid>"` URLs set to the aliases are only there to indicate that it would be impossible for the test case to _work_ using the traditional mirror configuration keys, but that it works _only_ by virtue of the `SourceMirror` plugin parsing the additional data set in the plugin configuration (`config:` section)

The API changes in the YAML, compared to before, are only the following:

* When defining `mirrors:`
  * The `mirror` definition now has an optional `kind:`, which can be used to refer to a plugin (defaulting to the default `SourceMirror` base plugin in the core)
  * The `mirror` definition now has an optional `config:` section, which can be parsed and validated by the plugin, in `Plugin.configure()` in the normal way.

Does this answer your questions ? I think you are misunderstanding this test case and I hope I've clarified it, otherwise I am perhaps misunderstanding the issue you are raising.

Reply:
Additional comment about _"needing to set mirror values for aliases"_ in general.

As I already mentioned, yes I am in favor of expanding this later in order to support more generic application of `SourceMirror`, perhaps allowing a project to just declare a single `SourceMirror` object once and allow that to handle all of the url translations.

That said, there is a lot to work out for that, e.g. we have priorities and lists etc in the API to consider.

Finally to answer _that_ question about needing to specify the aliases with at least some _values_ for the aliases in a given mirror, no it will not possibly work without specifying aliases in the patch's current state.

The reason for this is that the current, existing workflow, taking all mirrors and priorities etc into account, needs to run through `Project.get_alias_uris()`, this is where the URI fetch/track policies are considered and this is the code section which controls the priority order of _"each URI which shall be consulted for a given alias"_.

In order to extend `SourceMirror` to apply more generically, this whole concept of _iterating over URIs for a given alias_ needs to be rethought.

Note also, that the _same `SourceMirror`_ may be consulted _multiple times_ should it be given more than a single URL value for a given alias that it declares (which was of course, already the case before introducing this plugin).

Reply:
Just to make sure, did you see my other comment? https://github.com/apache/buildstream/pull/1890#discussion_r1509989058. I feel that you're answering in the wrong thread.

I agree with what you say here, but I want to change it before landing this.

> yes I am in favor of expanding this later in order to support more generic application of SourceMirror

> no it will not possibly work without specifying aliases in the patch's current state.

I'm afraid we can't reconcile these two statements. If we want to expand this later, we shouldn't make such a hard requirement. We need to leave ourselves some space to work with.

Reply:
I understand what you're saying now.

I would have rather expected:


```yaml
- name: freedesktop-sdk-raw-file-mirrors
  kind: gitlab_lfs_mirror
  aliases:
    cpan: https://gitlab.com/freedesktop-sdk/mirrors/{alias}/raw-files
    crates: https://gitlab.com/freedesktop-sdk/mirrors/{alias}/raw-files
    ftp_gnu_org: https://gitlab.com/freedesktop-sdk/mirrors/{alias}/raw-files
    ...
```

I.e. we're still expecting the existing mirror configuration to be valuable, and used in the default `SourceMirror` implementation (base class).

There are two paths I can see towards supporting this.

## Allow list-or-dict declaration of the aliases

In your example:

```yaml
- name: freedesktop-sdk-raw-file-mirrors
  kind: gitlab_lfs_mirror
  url: https://gitlab.com/
  project: freedesktop-sdk/mirrors/{alias}/raw-files
  aliases:
  - cpan
  - crates
  - ftp_gnu_org
  - github_files
```

This could be supported on the condition that the mirror is a user plugin, and would result in passing `None` values for the configured URI passed through `SourceMirror.translate_url()`.

## Support non aliased URIs

In your example:

```yaml
- name: freedesktop-sdk-raw-file-mirrors
  kind: gitlab_lfs_mirror
  config:
    url: https://gitlab.com/
    project: freedesktop-sdk/mirrors/{alias}/raw-files
    aliases:
    - cpan
    - crates
    - ftp_gnu_org
    - github_files
```

This will require restructuring the `Source.__do_fetch()` and `Source.__do_track()` loops and is a bit more involved.

I.e. we would want a mirror to be allowed to apply to _"any alias"_ and/or _"no alias"_.

For the _"no alias"_ case it is a bit more clear cut, afaics there are specific code blocks where we simply do not as the `Project` about the urls for aliases and directly invoke the plugin implementation.

### For plugins who want to support _"any alias"_

For the _"any alias"_ case, maybe we would want a `None` URL with a `SourceMirror` tuple to be returned in the list from `Project.get_alias_uris()` (allowing the project to continue to respect the user configuration of priorities).

Both cases will require that `Source.__clone_for_uri()` work with a `None` uri and/or a `None` alias.

The last part, for blanket mirrors to not need to specify the alias/[urls], it would make sense for `Source.translate_url()` to be allowed to _"not work"_ when given unrecognized aliases.

This could allow specification of multiple SourceMirror plugins which do things in different ways, and can bail out on aliases/urls which that plugin does not support, also it allows falling back to other mirrors or original project alias if the user has configured `SourceUriPolicy.ALL` and the mirror failed to produce a URI.

This is tricky to accomplish with the current code structure though. At fetch/track time it could work by raising an exception, allowing the loop to continue when the SourceMirror does:

```python
def translate_url(self, ... alias ...):

    if i_dont_know_this(alias):
        raise SourceMirrorUnhandled()
```

This would abort the plugin code calling into `Source.translate_url()` without breaking existing plugin code, as long as mirrors are never used in `Plugin.configure()` time and only at `Source.fetch()` and `Source.track()` time (we would only end up throwing away cloned Sources in the case that a URI could not be produced, and move on to the next mirror/uri).

Reply:
Tbf, I think that we can have all the needed flexibility (at least in the case where there is an alias) by making `SourceMirror._get_alias_uris()` public for plugins to implement. This allows the plugins to express whatever they can support.

As I said before, I think we could have this method just return the number of URLs it will provide, and have `SourceMirror.translate_url()` take an index rather than a "URL" which isn't exactly going to be a URL depending on what the source mirror is doing internally.

In the absence of an alias, I think we can just pass `None` as an alias. `Project.get_alias_uris()` currently allows the alias to be `None`, so it is more a question of forwarding this to mirrors, and making sure sources and source fetchers don't explode when given a substitution URL for a `None` alias. We may not do this now, but explicitly state that alias can be `None` in `SourceMirror._get_alias_uris()` (or whatever it will be called once it's made public).

The only question that remains is about a mirror plugin applying to more than one project: In the current state, a mirror is a property of a project, and `SourceMirror._get_alias_uris()` doesn't know about which project it is answering for. But I think this is a separate question.